packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
+  ) public view returns (bytes32[] memory) {
+    // Same as for the other lookup, preallocate and then trim for now
+    uint256 currIndex = 0;
+    bytes32[] memory identifiers = new bytes32[](maxIdentifiers);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      address trustedIssuer = trustedIssuers[i];
+      uint256 numIssuersForAddress = addressToIdentifiers[account][trustedIssuer].length;
+      for (uint256 j = 0; j < numIssuersForAddress; j = j.add(1)) {
+        // Iterate through the list of identifiers
+        if (currIndex < maxIdentifiers) {
+          bytes32 identifier = addressToIdentifiers[account][trustedIssuer][j];
+          // Check if the mapping was produced by a revoked signer
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation[] memory attestationsForIssuer = identifierToAddresses[identifier][trustedIssuer];
+          for (uint256 k = 0; k < attestationsForIssuer.length; k = k.add(1)) {
+            OwnershipAttestation memory attestation = attestationsForIssuer[k];
+            // (identifier, account, issuer) tuples should be unique
+            if (attestation.account == account && !revokedSigners[attestation.signer]) {
+              identifiers[currIndex] = identifier;
+              currIndex = currIndex.add(1);
+              break;
+            }
+          }
+        } else {
+          break;
+        }
+      }
+    }
+    if (currIndex < maxIdentifiers) {
+      // Allocate and fill properly-sized array
+      bytes32[] memory trimmedIdentifiers = new bytes32[](currIndex);
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
+    require(
+      msg.sender == account ||
+        msg.sender == issuer ||
+        getAccounts().attestationSignerToAccount(msg.sender) == issuer,
+      "User does not have permission to perform this action"
+    );
+    require(!revokedSigners[msg.sender], "User has been revoked ");
+    _;
+  }
+
+  /**
+   * @notice Validates the given attestation and signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @return Whether the signature is valid
```

Comment:
It additionally checks whether the signer is revoked, and if the signer is (currently) allowed as an attestation signer for the issuer.
```suggestion
   * @return Whether the signature is valid and the signer is authorized to make attestations.
```